elastic/kibana
# [Fleet] Support subfeatures privileges for Fleet APIs

## Summary

Depends on https://github.com/elastic/kibana/pull/178006

That PR add support for subfeatures privileges for Fleet APIs. Accessible behind a feature flag.

Renamed subfeature id to be consistent with other plugins.

## Will do in a follow up PR

Read agents should give access to agent policies, I will make a follow up PR for that, to avoid adding to much complexity to that one. 

It's possible that some API access need to change when implementing the UI, let discuss this in the following PRs

## Tests

- [x] unit test to the function that calculate auth for API.
- [x] API tests updated and added some tests for api with new subfeature roles

### Manual tests

You need to enable trial licence and the `subfeaturePrivileges` feature flag , than you create some custom fleet role (from the API, the UI is affected by this bug https://github.com/elastic/kibana/pull/178239)

Than you can try to reach the different fleet API and check if you have access or not.

Example 

```
curl -X PUT "<kibana_url>/api/security/role/fleet_test" -u elastic:changeme -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d'
{
  "metadata": {
    "version": 1
  },
  "elasticsearch": {
    "cluster": [ ],
    "indices": [ ]
  },
  "kibana": [
    {
      "base": [ ],
      "feature": {
       "fleetv2": [ "agents_read" ]
      },
      "spaces": [ "*" ]
    }
  ]
}
'
```

```
curl -X POST "<es_url>/_security/user/fleet_user?pretty" -u elastic:changeme -H 'Content-Type: application/json' -d'
{
  "password" : "changeme",
  "roles" : [ "fleet_test" ],
  "full_name" : "Jack Nicholson",
  "email" : "jacknich@example.com"
}
'
```
 should be able to access agents
```
curl <kibana_url>/api/fleet/agents -u fleet_user:changeme
```
 should not be able to access package policies
```
curl <kibana_url>/api/fleet/package_policies -u fleet_user:changeme
```
